node: property string list helper

rmsyn opened the pull request:
Adds a helper function to convert a `NodeProperty` value into a list of `&str`.

rmsyn commented on src/node.rs:
```diff
@@ -550,6 +550,18 @@ impl<'a> NodeProperty<'a> {
+    pub fn as_str_list(self) -> impl Iterator<Item = &'a str> {
+        let mut start = 0;
+
+        core::iter::from_fn(move || {
```
This stops after the first value, and stops updating the `stop` variable.

Not sure if there is an actual way to get this to work as intended.

rmsyn replied:
This stops after the first value, and stops updating the `start` variable.

Not sure if there is an actual way to get this to work as intended. I think it is failing because the `value` slice is consumed on the first call, becoming zero-length on the second iterator call.

rmsyn commented on src/tests.rs:
```diff
@@ -271,3 +271,17 @@ fn interrupt_cells() {
+    let expected = ["sifive,test1", "sifive,test0", "syscon"];
+    let compat = test.property("compatible").unwrap();
+
+    assert_eq!(compat.as_str_list().count(), expected.len());
```
Added a failing assertion to check the expected length.

repnop commented on src/node.rs:
```diff
@@ -550,6 +550,12 @@ impl<'a> NodeProperty<'a> {
     }
 
+    /// Attempts to parse the property value as a list of [`&str`].
+    pub fn as_str_list(self) -> Option<impl Iterator<Item = &'a str> + 'a> {
```
A better method name would be ideal, something akin to `iter_strs` perhaps? (the [API Naming Guidelines](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv) have a specific meaning for methods beginning with `as_`, so I'd like to try and keep things relatively consistent with that)

rmsyn replied:
Sounds good to me, I'll change it.

repnop commented on src/node.rs:
```diff
@@ -550,6 +550,12 @@ impl<'a> NodeProperty<'a> {
 
+    /// Attempts to parse the property value as a list of [`&str`].
+    pub fn as_str_list(self) -> Option<impl Iterator<Item = &'a str> + 'a> {
+        let mut s = self.as_str()?.split('\0');
+        Some(core::iter::from_fn(move || s.next()))
```
This could be simplified down to `Some(self.as_str()?.split('\0'))`, I believe

rmsyn replied:
> This could be simplified down to Some(self.as_str()?.split('\0')), I believe

No, because the `?` operator cannot be used in a function that doesn't return an `Option/Result` type.

rmsyn replied:
> This could be simplified down to Some(self.as_str()?.split('\0')), I believe

The `Split` iterator returned does not match the `impl Iterator<Item = &'a str> + 'a` return type. I tried this before, and using the `core::iter::from_fn` was the cleanest implementation that works, and stays consistent with the rest of the library.

rmsyn replied:
The reason I implemented it this way is because removing the `Option` wrapper from the return type requires separating the iterator creation.

I force-pushed a commit with the version of this function that returns a bare `impl Iterator` type.

repnop reviewed (changes requested):
A few small quibbles

rmsyn committed:
node: property string list helper

Adds a helper function to convert a `NodeProperty` value into a list of
`&str`.
